### PR TITLE
Restructure scripts to conform with other demos; add Jenkins support;…

### PR DESCRIPTION
**Pull request by jramsdale**
… update documentation

### CODE REVIEW

**Comment by chrislovecnm** [comment]
please test this

**Comment by chrislovecnm** [comment]
please test this

File: scripts/create.sh [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+source "$ROOT/scripts/common.sh"
+
+VERSION=$(cat "${ROOT}/VERSION")
+
+check_dependency_installed gcloud
+check_dependency_installed terraform
+
+# Enable required GCP services
+get_project
+enable_project_api ${PROJECT} compute.googleapis.com
```
Comment (geojaz) [inline scripts/create.sh]: Need to do: `enable_project_api "${PROJECT}" container.googleapis.com`

Comment (jramsdale) [reply scripts/create.sh]: That's a change that should be made, but are you sure it's to blame for the validation failure? If so, can you explain for me?

**Comment by jramsdale** [comment]
test this please

File: README.md [hunk 2]
```diff
@@ -74,44 +51,39 @@ Kubernetes directly.
...
 `http://[ip]:8080/factorial/` and
 
 `http://[ip]:8080/prime/`
 
-examples would look like:
+Examples of use would look like:
 
 ```console
-curl http://[ip]:8080/prime/10
+curl http://35.227.149.80:8080/prime/10
```
Comment (geojaz) [inline README.md]: unless there is a reason to specify a static IP here, I would suggest a place holder or var. ie: `curl "http://${ip}:8080/prime/10"

File: README.md [hunk 3]
```diff
@@ -74,44 +51,39 @@ Kubernetes directly.
...
 The sum of all primes less than 10 is 17
 
-curl http://[ip]:8080/factorial/10
+curl http://35.227.149.80:8080/factorial/10
```
Comment (geojaz) [inline README.md]: as above

File: README.md [hunk 4]
```diff
@@ -123,62 +95,46 @@ gcloud auth application-default login
 
 ### Configure gcloud settings
 
+The scripts that manage the infrastructure and deployments for this demo attempt to initialize themselves with values from the `gcloud` config defaults for `compute/zone`, and `core/project`.
```
Comment (geojaz) [inline README.md]: Perhaps either force folks to do this config by calling a make target that verifies this is all set before calling make validate OR call it out more directly. IE: you cannot continue if these condition haven't been met. I lean towards automating and letting that code serve as forceful documentation.

File: README.md [hunk 5]
```diff
@@ -187,75 +143,76 @@ At the root of this repository, execute `setup.sh`. It will:
...
-Validating these three deployments is done by simply executing `validate.sh`
-from the root of the directory.
-A successful output will contain:
+Validating these three deployments is done by executing:
+
+```console
+make validate
+```
+
+A successful output will look like this:
```
Comment (geojaz) [inline README.md]: ... wil look something like this: (with variables instead of hardcoded ips)

File: README.md [hunk 6]
```diff
@@ -187,75 +143,76 @@ At the root of this repository, execute `setup.sh`. It will:
...
 
+Of course, the IP addresses will likely differ for your deployment.
+
 ## Load Testing
 
-In a new console window, execute (using ip-3 from above)
+In a new console window, execute the following, replacing `[IP_ADDRESS]` with the IP address and port from your validation output from the previous step. Note that the Kubernetes deployment runs on port `80`, while the other two deployments run on port `8080`:
 ```console
-ab -c 120 -t 60  http://[ip-3]/prime/10000
+ab -c 120 -t 60  http://[IP_ADDRESS]/prime/10000
```
Comment (geojaz) [inline README.md]: those changes to parameterize the IP will make this better.

File: README.md [hunk 7]
```diff
@@ -187,75 +143,76 @@ At the root of this repository, execute `setup.sh`. It will:
...
+![screenshot](./images/ab_load-test-1.png)
 
-A good mechanism for scaling this type of service is via horizontal scaling
-(adding additional instances behind a load balancer). For either of the
-first two deployments (Debian or COS), this would take several steps:
+A good mechanism for scaling this type of service is via horizontal scaling (adding additional instances behind a load balancer). For either of the first two deployments (Debian or COS), this would take several steps:
 
 1.  Create a load balancer.
-1.  Create additional instances.
+1.  Create additional application instances.
```
Comment (geojaz) [inline README.md]: is this scaling the replicas in deployment or additional applications?

File: README.md [hunk 8]
```diff
@@ -187,75 +143,76 @@ At the root of this repository, execute `setup.sh`. It will:
...
+This is an involved process and is out of scope for this demonstration.
 
 For the third (Kubernetes) deployment the process is far easier:
 
 ```console
 kubectl scale --replicas 3 deployment/prime-server
 ```
 
-After allowing 30 seconds for the replicas to initialize, re-run.
+After allowing 30 seconds for the replicas to initialize, re-run the load test:
```
Comment (geojaz) [inline README.md]: can we leverage a readiness check or is there something more manual that is needed?

File: README.md (hunk 2 above)
Comment (jramsdale) [reply README.md]: So, I disagree. I think console output should be representative. When example commands are given I do agree, however.

File: README.md (hunk 3 above)
Comment (jramsdale) [reply README.md]: Ultimately, as the reviewer you get to decide. Lemme know if you find my argument convincing or if you'd like to discuss more!

File: README.md (hunk 4 above)
Comment (jramsdale) [reply README.md]: Ultimately the task was to add this existing project to Jenkins and I was trying not to expand the scope too much. I agree that it would be a nice improvement--I think the question is whether it needs to be part of this particular PR. Thoughts?

File: README.md (hunk 5 above)
Comment (jramsdale) [reply README.md]: There's a comment to that effect underneath the console output. Are you suggested merge them into one comment above?

File: README.md (hunk 6 above)
Comment (jramsdale) [reply README.md]: Not sure I follow?

File: README.md [hunk 9]
```diff
@@ -30,38 +20,25 @@ different containerized cloud infrastructure deployments.
 
 ## Introduction
 
-Containers are superior to conventional deployment mechanism in that they are:
-
-1. Isolated:
-  - Applications have their own libraries, no conflicts will arise from
-  different libraries in other applications.
-
+Containers are superior to conventional deployment mechanisms in that they are:
```
Comment (geojaz) [inline README.md]: Can we rephrase this to something like: There are many advantages to building your deployment pipeline around containers... It's a nit, but saying that containers are always superior seems like a hard claim to back up and is a little far reaching, IMO

File: README.md [hunk 10]
```diff
@@ -30,38 +20,25 @@ different containerized cloud infrastructure deployments.
...
-This project demonstrates migrating a simple Python application named Prime-flask:
+1.  A containerized version deployed on [Container-Optimized OS (COS)](https://cloud.google.com/container-optimized-os/)
 
-1.  A legacy deployment (Debian VM) where Prime-flask is deployed as the only
-    application much like a traditional application was run in an
-    on-premises datacenter.
-1.  Second, a containerized version is deployed on Container-Optimized OS (COS).
-1.  Finally, a Kubernetes deployment where Prime-flask runs as a deployment
-    behind a load-balancer cluster in Google Kubernetes Engine.
+1.  A [Kubernetes](https://kubernetes.io/) deployment where `Prime-flask` runs as a deployment behind a load-balancer cluster in Google [Kubernetes Engine](https://cloud.google.com/kubernetes-engine/)
```
Comment (geojaz) [inline README.md]: s/runs as a deployment behind/is exposed via a load balancer and deployed in Kubernetes Engine.

File: README.md [hunk 11]
```diff
@@ -74,44 +51,39 @@ Kubernetes directly.
...
-  and get 300 dollars of free credit on your new account.
+[Kubernetes Engine v1.10.0 or later](https://cloud.google.com/kubernetes-engine/docs/quickstart#before-you-begin) service enabled
+* If you do not have a Google Cloud Platform account you can sign up [here](https://cloud.google.com) and get 300 dollars of free credit on your new account.
 * [Google Cloud SDK (200.0.0 or later)](https://cloud.google.com/sdk/downloads)
 * [ApacheBench](https://httpd.apache.org/docs/2.4/programs/ab.html)
 * [HashiCorp Terraform (>= v0.11.7)](https://www.terraform.io/downloads.html)
 * [gcloud](https://cloud.google.com/sdk/gcloud/)
 * [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
-
+* [Terraform](terraform.io)
```
Comment (geojaz) [inline README.md]: 83 seems to address terraform, I think we can delete this line

File: README.md [hunk 12]
```diff
@@ -123,62 +95,46 @@ gcloud auth application-default login
...
-
-At the root of this repository, execute `setup.sh`. It will:
-1.  Package the deployable Prime-flask application.
-1.  Create the container image and push it to the private image repository
-    in your project.
-1.  Generate variable defaults for the Terraform execution.
+This will:
+1.  Package the deployable `Prime-flask` application.
+1.  Create the container image and push it to the private [Container Registry (GCR)](https://cloud.google.com/container-registry/) for your project.
+1.  Generate variable defaults for the [Terraform](terraform.io) execution.
```
Comment (geojaz) [inline README.md]: generate appropriate configuration for terraform.

File: README.md (hunk 6 above)
Comment (geojaz) [reply README.md]: Sorry, I think I was talking to myself on this one. But... let's use a legal bash variable `$IP_ADDRESS` instead of `[IP_ADDRESS]`.

File: README.md [hunk 13]
```diff
@@ -187,75 +143,76 @@ At the root of this repository, execute `setup.sh`. It will:
...
 
-This can be confirmed by reviewing the output from the `ab` command. A
-"Failed requests" value of more than 0 means that the server couldn't
-respond successfully for this amount of load.
+![screenshot](./images/ab_load-test-1.png)
 
-A good mechanism for scaling this type of service is via horizontal scaling
-(adding additional instances behind a load balancer). For either of the
-first two deployments (Debian or COS), this would take several steps:
+A good mechanism for scaling this type of service is via horizontal scaling (adding additional instances behind a load balancer). For either of the first two deployments (Debian or COS), this would take several steps:
```
Comment (geojaz) [inline README.md]: One way to ensure that your system has capacity to handle this type of traffic is by scaling up. In this case, we would want to [scale our service horizontally](https://en.wikipedia.org/wiki/Scalability#Horizontal_and_vertical_scaling).

In our Debian and COS architectures, horizontal scaling would include: 
1. Creating a load balancer
1. Spinning up additional instances
1. Registering them with the load balancer

File: README.md (hunk 2 above)
Comment (geojaz) [reply README.md]: I'm picking up what you're putting down. I'm cool with this. but in this case, please use real bash syntax in 56 and 58 for vars (same comment throughout)

Comment (jramsdale) [reply README.md]: Per off-line conversation, we're agreed on `<ip>`

File: README.md (hunk 6 above)
Comment (jramsdale) [reply README.md]: Per above, changed to `<>`

File: README.md (hunk 5 above)
Comment (geojaz) [reply README.md]: let's keep as is as per our discussion

File: README.md (hunk 11 above)
Comment (jramsdale) [reply README.md]: Yup

File: README.md (hunk 9 above)
Comment (jramsdale) [reply README.md]: 100% agreed. Probably an overly-light touch for me to leave that in.

File: README.md (hunk 10 above)
Comment (jramsdale) [reply README.md]: k

File: README.md (hunk 12 above)
Comment (jramsdale) [reply README.md]: k

File: README.md (hunk 13 above)
Comment (jramsdale) [reply README.md]: 👍

**Comment by chrislovecnm** [comment]
We are gonna need to squash these commits into units of work.

**Comment by cxhercules** [comment]
Commits squashed 👍
